multichain-aggregator/multichain-aggregator-logic/src/repository/addresses.rs

```diff
@@ -55,25 +58,39 @@ pub async fn list<C>(
 where
     C: ConnectionTrait,
 {
-    let addresses = Entity::find()
+    // Materialize addresses CTE when searching by contract_name.
+    // Otherwise, query planner chooses a suboptimal plan.
+    let is_cte_materialized = query.is_some();
+    let addresses_cte = CommonTableExpression::new()
+        .query(
+            QuerySelect::query(&mut Entity::find())
+                .apply_if(query, |q, query| {
+                    let ts_query = prepare_ts_query(&query);
+                    q.and_where(Expr::cust_with_expr(
+                        "to_tsvector('english', contract_name) @@ to_tsquery($1)",
+                        ts_query,
+                    ));
+                })
+                .to_owned(),
+        )
+        .materialized(is_cte_materialized)
+        .table_name(Alias::new("addresses"))
+        .to_owned();
+
+    let base_select = Query::select()
+        .column(ColumnRef::Asterisk)
+        .from(Alias::new("addresses")) // NOTE: this is the CTE reference
```

Comment:
maybe it's more clear to make an alias before `let addresses_cte` and use it in line 77 and here?